### PR TITLE
pep612: add semanal for paramspec

### PR DESCRIPTION
**Pull request by hauntsaninja**
Linking #8645

Taking Jukka's suggestion from #9250 and splitting up the PR so we can merge and review a piece at a time. If you've read what's been pushed to #9250, this is the first bit of it, except I've changed mypy_extensions to typing_extensions since PEP 612 has been accepted.

### CODE REVIEW

File: mypy/semanal.py [hunk 1]
```diff
@@ -3008,6 +3018,41 @@ def process_typevar_parameters(self, args: List[Expression],
...
 
+    def process_paramspec_declaration(self, s: AssignmentStmt) -> bool:
+        """Checks if s declares a ParamSpec; if yes, store it in symbol table.
+
+        Return True if this looks like a parameter specification declaration (but maybe
+        with errors), otherwise return False.
+
+        In the future, ParamSpec may accept bounds and variance arguments, in which
+        case more aggressive sharing of code with process_typevar_declaration should be pursued.
+
```
Comment (JukkaL) [inline mypy/semanal.py:3029]: Nit: redundant empty line at the end of docstring.

File: test-data/unit/semanal-errors.test [hunk 2]
```diff
@@ -1419,3 +1419,12 @@ def g() -> None:
                           # N: (Hint: Use "T" in function signature to bind "T" inside a function)
 [builtins fixtures/dict.pyi]
 [out]
+
+[case testParamSpec]
+from typing_extensions import ParamSpec
+
+TParams = ParamSpec('TParams')
+TP = ParamSpec('?')  # E: String argument 1 '?' to ParamSpec(...) does not match variable name 'TP'
+TP2: int = ParamSpec('TP2')  # E: Cannot declare the type of a parameter specification
```
Comment (JukkaL) [inline test-data/unit/semanal-errors.test:1428]: Optionally, you can add a test case to `test-data/unit/semanal-symtable.test` and/or `test-data/unit/semanal-types.test` (these are not super high signal test cases, but on the other hand they don't require a lot of maintenance).

**Review by JukkaL — APPROVED** [review]
Thanks! Looks good, just left a few minor comments. Feel free to merge without waiting for another review round. 

(Feel free to ping me on Gitter or send an email if you have a new PR up -- I want to prioritize reviewing PEP 612 related PRs, since this is an important new feature.)

**Comment by hauntsaninja** [comment]
Thanks, made the changes. Had to add two visitors to get the test working. Also got around to recognising `typing.ParamSpec`, since otherwise treetransform outputs a lot of stuff coming from `typing_extensions`.

**Pull request merged**